lsp-mode.el

```diff
@@ -161,29 +194,48 @@ Optional arguments:
-                       :send-async #'lsp--stdio-send-async
-                       :new-connection (lsp--make-tcp-connection
+    `(progn
+       (lsp-define-whitelist-enable ,name ,get-root)
+       (lsp-define-whitelist-disable ,name ,get-root)
+       (lsp-define-interactive-client ,name)
+
+       (defun ,enable ()
+         ,docstring
+         (interactive)
+         (let ((client (make-lsp--client
+                        :language-id (if ,language-id-fn ,language-id-fn #'(lambda (_) ,language-id))
+                        :send-sync #'lsp--stdio-send-sync
+                        :send-async #'lsp--stdio-send-async
+                        :new-connection (lsp--make-tcp-connection
                                          ,(symbol-name name)
                                          ,command
                                          ,command-fn
                                          ,host ,port)
-                       :get-root ,get-root
-                       :ignore-regexps ,ignore-regexps)))
-         (unless lsp-mode
-           ,(when initialize
-              `(funcall ,initialize client))
-           (let ((root (funcall (lsp--client-get-root client))))
-             (if (lsp--should-start-p root)
-               (progn
-                 (lsp-mode 1)
-                 (lsp--start client ,extra-init-params))
-               (message "Not initializing project %s" root))))))))
+                        :get-root ,get-root
+                        :ignore-regexps ,ignore-regexps)))
+           (lsp--start-client-maybe client :extra-init-params ,extra-init-params :initialize ,initialize))))))
+
+(cl-defun lsp--start-client-maybe (client
+                                 &key extra-init-params initialize)
+ "Start a client if lsp-mode is currently inactive, and the whitelist/blacklist allows it."
+            (unless lsp-mode
+              (when initialize
+                 (funcall initialize client))
+              (let ((root (funcall (lsp--client-get-root client))))
+                (if (lsp--should-start-p root)
+                    (progn
+                      (lsp-mode 1)
+                      (lsp--start client extra-init-params))
+                  (if (and ask lsp-enable-whitelist-ask)
+                      (let ((question (format-message "Add %s to lsp-project-whitelist? " root)))
+                        (if (y-or-n-p question) ;; cannot use when , side effects problem
+                            (progn (message "About to customize lsp-project-whitelist")
+                                   (customize-save-variable 'lsp-project-whitelist (add-to-list 'lsp-project-whitelist root)))
+                          (setq lsp--project-whitelist-temp (add-to-list 'lsp--project-whitelist-temp root)))
+                        ;; start regardless, this is called interactively
+                        (lsp-mode 1)
+                        (lsp--start client extra-init-params))
```

Comment:
The client is started whether the user selects `y` or `n` in the question. I would suspect that the language server is not started if the I select `n` to white-listing.

Reply:
No. The key feature of this PR is that we end up with `lsp-MAJOR-MODE-enable -hook` and `lsp-MAJOR-MODE-enable`.

The first one gets added to the `MAJOR-MODE` hook, and sets the `ask` parameter to nil.

The second one is called via `M-x` (i.e. interactively) and sets the `ask` parameter to `t`.

The logic is that `lsp-MAJOR-MODE-enable` will *always* start a session.  The only difference is that it may prompt to *also* add the project root to the whitelist, for future sessions.

Reply:
But who would call the second interactively, if he already has enabled the hook? 
And if you always want to call this interactively, you don't need the whitelist, as you are the one who decides when to start lsp.

Reply:
I have a mixed case. Some projects I always want to start, others to only start by request.

Especially as the haskell lsp-server uses quite a bit of resources, so if you just want to look at code for a project while working on another, you do not necessarily want to start a server.

Reply:
Hmm, I understand. Then I would propose. You remove the asking stuff of the interactive function, but preserve the interactive function. If you have a project that you always want to start by request, you use the interactive version. If you decide, for this project the language server should always be started you call `lsp-whitelist-project-add`.